core/src/test/scala/integration/kafka/api/ProducerIdExpirationTest.scala

```diff
@@ -0,0 +1,176 @@
+  def testProducerIdExpirationWithNoTransactions(quorum: String): Unit = {
+    producer = TestUtils.createProducer(bootstrapServers(), enableIdempotence = true)
+
+    // Send records to populate producer state cache.
+    producer.send(new ProducerRecord(topic1, 0, null, "key".getBytes, "value".getBytes))
+    producer.flush()
+
+    // Ensure producer IDs are added.
+    ensureConsistentKRaftMetadata()
+    assertEquals(1, producerState.size)
+
+    // Wait for the producer ID to expire
+    TestUtils.waitUntilTrue(() => producerState.isEmpty, "Producer ID did not expire.")
+
+    // Send more records to send producer ID back to brokers.
+    producer.send(new ProducerRecord(topic1, 0, null, "key".getBytes, "value".getBytes))
+    producer.flush()
+
+    // Producer IDs should repopulate.
+    assertEquals(1, producerState.size)
+  }
+
+  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumName)
+  @ValueSource(strings = Array("zk", "kraft"))
+  def testTransactionAfterTransactionIdExpiresButProducerIdRemains(quorum: String): Unit = {
+    producer = TestUtils.createTransactionalProducer("transactionalProducer", brokers)
+    producer.initTransactions()
+
+    // Start and then abort a transaction to allow the producer ID to expire
+    producer.beginTransaction()
+    producer.send(TestUtils.producerRecordWithExpectedTransactionStatus(topic1, 0, "2", "2", willBeCommitted = false))
+    producer.flush()
+
+    // Ensure producer IDs are added.
+    assertEquals(1, producerState.size)
+
+    producer.abortTransaction()
+
+    // Wait for the transactional ID to expire
+    Thread.sleep(3000)
+
+    // Confirm the transactional IDs expired
+    val txnDescribeResult = admin.describeTransactions(Collections.singletonList("transactionalProducer")).description("transactionalProducer")
+    org.apache.kafka.test.TestUtils.assertFutureThrows(txnDescribeResult, classOf[TransactionalIdNotFoundException])
+
+    // Producer IDs should be retained.
+    assertEquals(1, producerState.size)
+
+    // Start a new transaction and attempt to send, which will trigger an AddPartitionsToTxnRequest, which will fail due to the expired transactional ID
+    producer.beginTransaction()
+    val failedFuture = producer.send(TestUtils.producerRecordWithExpectedTransactionStatus(topic1, 3, "1", "1", willBeCommitted = false))
+    TestUtils.waitUntilTrue(() => failedFuture.isDone, "Producer future never completed.")
+
+    org.apache.kafka.test.TestUtils.assertFutureThrows(failedFuture, classOf[InvalidPidMappingException])
+    producer.abortTransaction()
+
+    producer.beginTransaction()
+    producer.send(TestUtils.producerRecordWithExpectedTransactionStatus(topic1, 2, "4", "4", willBeCommitted = true))
+    producer.send(TestUtils.producerRecordWithExpectedTransactionStatus(topic1, 3, "3", "3", willBeCommitted = true))
+    producer.commitTransaction()
+
+    // Producer IDs should be retained.
+    assertEquals(1, producerState.size)
```

Comment:
For my understanding, this assertion verifies that the producer ID created by the first transaction is still there, right? As the last transaction produces to partition 2 and 3, should we verify them as well?

Reply:
If it makes it less confusing, I can just produce to one partition. Not sure we get much out of checking multiple and it will make the helper more complicated. Let me know if you think it provides value.

Reply:
I am trying to understand the value provided by this assertion. There is also a risk for the producer id to be expired by the time we reach it btw because we don't "refresh it" for partition 0 with the last transaction, right?

Reply:
Yeah -- this might be a mistake. I can just produce to partition 0 there.